### ISSUE TITLE
Soft lockup when running pkt-gen

### ISSUE THREAD

**Issue by zdrost** [luigirizzo/netmap#383]
After changing from commit [3947b21](https://github.com/luigirizzo/netmap/commit/3947b2174f13fd54ae6db5e4831846a1811cd8c3) (May 5) to [4fc03d3](https://github.com/luigirizzo/netmap/commit/4fc03d309977b6d7350434767c136d1cc6a98667) (October 10), I've been getting these soft lockups while running pkt-gen over Netmap:

`Message from syslogd@testhost at Oct 11 16:53:24 ...
 kernel:NMI watchdog: BUG: soft lockup - CPU#0 stuck for 22s! [pkt-gen:29061]`

After that, machine is completely unresponsive and hard reset is the only way to get it back. This is quite burdensome since it's a remote machine. 

Configuration is as follows: OS is a CentOS 7 (kernel version: 3.10.0-514.el7.x86_64). Machine has a dual port NIC (Intel 82599ES 10Gbps SFI/SFP+) with the ports connected directly one to the other with a fiber. Netmap has been compiled against the 3.10.0-514.16.1.el7.x86_64 headers and the driver used is **ixgbe**.

Basically, I used one NIC port/interface to send data with pkt-gen and the other one to receive data with pkt-gen. Using the old commit [3947b21](https://github.com/luigirizzo/netmap/commit/3947b2174f13fd54ae6db5e4831846a1811cd8c3) this was working fine on numerous occasions. But with the newer commit, I'm getting these soft lockups.

Let me know if you need some other information.

**Comment by giuseppelettieri** [comment]
May you please upgrade to the latest master? There was a bug exposed by the Intel driver update and fixed in  commit 1597e0c.

**Comment by zdrost** [comment]
I've upgraded to the latest master and repeatedly tested both pkt-gen and some custom developed programs. They seem to be stable. I've probably landed on the wrong commit.

However, I'd like to make one more test on another machine (RHEL 7 with 3.10.0-693.2.2.el7.x86_64 kernel version, NIC is the same) where we first discovered this issue. That machine has a significantly more traffic incoming. Although the first test with latest master seems promising, let me give you the feedback in the next couple of days so that you can close this issue.

Thank you very much!

**Comment by andrewbonney** [comment]
I've just encountered the same soft lock up when closing a program using netmap on multiple hosts with a 10G Intel card (ixgbe). This is following a move from netmap commit https://github.com/luigirizzo/netmap/commit/56022755d9eeb6e7062fc3efce80aa476c981910 to https://github.com/luigirizzo/netmap/commit/866df6bc4251aca71ee6990ca8e587bb33d42e86 and from ixgbe 5.2.4 to 5.3.3.

I should have a little more time to look into it tomorrow, but hopefully this narrows things down if this is the same issue as reported here.

**Comment by giuseppelettieri** [comment]
A stack trace would be useful, if possible.

**Comment by andrewbonney** [comment]
The following comes from the kernel log following the lock up using netmap (866df6b) and ixgbe 5.3.3. I've just retried the same thing but reverting to ixgbe 5.2.4 with the same version of netmap and haven't experienced the issue.

It may be that this is a different issue to the one originally reported, but that it results in the same symptom. I'll do some more tests on the differences between the two versions as it's always possible there's been an error in our patching between 5.2.4 and 5.3.3.

```
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779223] NMI watchdog: BUG: soft lockup - CPU#1 stuck for 23s! [netmap-gateway:1465]
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779226] Modules linked in: xt_CHECKSUM iptable_mangle ipt_MASQUERADE nf_nat_masquerade_ipv4 iptable_nat nf_conntrack_ipv4 nf_defrag_ipv4 nf_nat_ipv4 userclock(OE) nf_nat nf_conntrack xt_tcpudp rpcsec_gss_krb5 auth_rpcgss bridge nfsv4 stp llc nfs iptable_filter ip_tables lockd x_tables grace fscache zram lz4_compress gpio_ich intel_powerclamp coretemp kvm irqbypass dcdbas 8250_fintek acpi_power_meter shpchp i7core_edac edac_core lpc_ich ipmi_ssif mac_hid ixgbe_bbcrd(OE) vxlan ip6_udp_tunnel udp_tunnel ptp pps_core dca sunrpc ipmi_watchdog ipmi_si ipmi_devintf ipmi_msghandler netmap(OE) autofs4 ses enclosure pata_acpi mpt3sas bnx2 raid_class scsi_transport_sas fjes
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779262] CPU: 1 PID: 1465 Comm: netmap-gateway Tainted: G           OE   4.4.0-98-generic #121-Ubuntu
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779263] Hardware name: Dell Inc. PowerEdge R310/0P229K, BIOS 1.1.5 05/24/2010
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779265] task: ffff88022eedd940 ti: ffff8800b64b8000 task.ti: ffff8800b64b8000
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779266] RIP: 0010:[<ffffffffc016844b>]  [<ffffffffc016844b>] ixgbe_clean_tx_ring+0xab/0x1b0 [ixgbe_bbcrd]
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779280] RSP: 0018:ffff8800b64bbb60  EFLAGS: 00000286
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779281] RAX: 0000000000000d1f RBX: ffffc900026145d0 RCX: 0000000000001000
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779282] RDX: 0000000000000000 RSI: 0000000000000000 RDI: ffff8802353b4098
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779283] RBP: ffff8800b64bbb90 R08: 0000000000000000 R09: 0000000000000d1f
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779284] R10: 0000000000000020 R11: 0000000000000000 R12: ffff88020cd4d1f0
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779286] R13: 0000000000000d1e R14: ffff8800aa4ab980 R15: 0000000000000000
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779287] FS:  00007fde0f614740(0000) GS:ffff880236e40000(0000) knlGS:0000000000000000
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779288] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779290] CR2: 00007fddea16e00c CR3: 00000002347d6000 CR4: 00000000000006e0
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779291] Stack:
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779292]  ffff8800b64bbb70 ffff8802334a0840 0000000000000001 ffff8802334a0a88
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779294]  ffff8802334a0000 ffff8802334a0f80 ffff8800b64bbbd8 ffffffffc016ecb1
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779296]  ffff8802334a1580 000000032eedd940 ffff8802334a0840 ffff8802334a0a50
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779297] Call Trace:
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779304]  [<ffffffffc016ecb1>] ixgbe_down+0x311/0x450 [ixgbe_bbcrd]
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779310]  [<ffffffffc016eec7>] ixgbe_reinit_locked+0x57/0xc0 [ixgbe_bbcrd]
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779316]  [<ffffffffc0173557>] ixgbe_do_reset+0x27/0x30 [ixgbe_bbcrd]
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779321]  [<ffffffffc017371b>] ixgbe_set_features+0x1bb/0x1e0 [ixgbe_bbcrd]
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779326]  [<ffffffff8173c237>] __netdev_update_features+0x247/0x760
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779329]  [<ffffffff81349e18>] ? security_capable+0x48/0x60
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779331]  [<ffffffff8173d58e>] __ethtool_set_flags+0x9e/0xb0
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779333]  [<ffffffff8173ff41>] dev_ethtool+0x1071/0x1f20
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779338]  [<ffffffff811a027b>] ? release_pages+0xdb/0x2b0
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779340]  [<ffffffff8174e355>] ? __rtnl_unlock+0x15/0x20
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779342]  [<ffffffff8173ce41>] ? netdev_run_todo+0x61/0x320
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779344]  [<ffffffff81752672>] dev_ioctl+0x182/0x590
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779347]  [<ffffffff811c70e2>] ? unmap_region+0xe2/0x120
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779351]  [<ffffffff817194b2>] sock_do_ioctl+0x42/0x50
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779353]  [<ffffffff817199b2>] sock_ioctl+0x1d2/0x290
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779356]  [<ffffffff812245cf>] do_vfs_ioctl+0x29f/0x490
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779358]  [<ffffffff811c957b>] ? do_munmap+0x34b/0x460
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779360]  [<ffffffff81224839>] SyS_ioctl+0x79/0x90
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779364]  [<ffffffff818446b2>] entry_SYSCALL_64_fastpath+0x16/0x71
Nov  9 09:24:34 ap-r310-0 kernel: [59809.779366] Code: 4c 8b 3b 41 0f b7 4e 4c 49 c1 e4 04 4d 03 66 20 4d 39 e7 75 0a e9 08 01 00 00 4d 39 fc 74 69 41 8d 45 01 48 83 c3 30 49 83 c4 10 <66> 39 c8 0f 84 c0 00 00 00 45 8d 4d 02 41 89 c5 8b 53 28 85 d2 
```

**Comment by giuseppelettieri** [comment]
Hi @andrewbonney, may you please try the latest master?

**Comment by andrewbonney** [comment]
Hi @giuseppelettieri. I gave this a try and still encountered the lock up. I'm going to try stripping back some of our own code patches in case these are affecting things. I'll post an update as soon as I've had chance to do this. Thanks

**Comment by zdrost** [comment]
Just to finally give the feedback from my side... I haven't encountered the soft lockup since changing to latest master on multiple hosts with various traffic loads. 

Excuse me for the long delay. Again, thanks a lot for the support.

**Comment by giuseppelettieri** [comment]
Thanks @zdrost.

@andrewbonney, please open a different issue for your case, if needed.
